netlogo-gui/src/main/editor/EditorMenu.scala

```diff
@@ -0,0 +1,9 @@
+// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo
+
+package org.nlogo.editor
+
+import javax.swing.Action
+
+trait EditorMenu {
+  def offerAction(action: Action): Unit
```

Comment:
Here are a two cases:

* Some component is holding a bunch of actions and doesn't know which one of them are menu entries, so it offers all of them to the menu to figure out what to present
* A menu is offered an action that doesn't correspond to its function (we currently split based on the menu category in `MenuBar`, but we could also check this in the menus themselves).

So the whole point of this massive menu refactoring (which is probably the majority of the code in this PR) is actually to separate different concerns. Menu's shouldn't know or care about *how* to perform the actions they contain; items which have a collection of actions (editor and tabs in this case) shouldn't know or care about *whether* an action will be placed in the menu; actions shouldn't care about *where* in the menu they are presented (and indeed, shouldn't really care whether they're presented at all, since the action doesn't know the state of the whole program and menu actions disappear for reasons unrelated to the action).

So by introducing the possibility that a menu may or may not add an action, we've moved the concern about whether an action is appropriate to present to the user to the two parties which know the most about the action: the menu and the action itself. If the action container knows that a menu item isn't appropriate (as Tabs sometimes does) it can use `revoke` without needing to worry about whether the menu is presenting the action or not.